geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/reader/CachingSimpleFeatureArrowFileReader.scala

```diff
@@ -0,0 +1,168 @@
+/***********************************************************************
+* Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.arrow.io.reader
+
+import java.io.{Closeable, InputStream}
+import java.nio.channels.{Channels, ReadableByteChannel}
+
+import org.apache.arrow.memory.BufferAllocator
+import org.apache.arrow.vector.complex.NullableMapVector
+import org.apache.arrow.vector.file.ReadChannel
+import org.apache.arrow.vector.schema.ArrowRecordBatch
+import org.apache.arrow.vector.stream.{ArrowStreamReader, MessageSerializer}
+import org.apache.arrow.vector.{VectorLoader, VectorSchemaRoot}
+import org.locationtech.geomesa.arrow.features.ArrowSimpleFeature
+import org.locationtech.geomesa.arrow.io.SimpleFeatureArrowFileReader.{SkipIndicator, VectorToIterator}
+import org.locationtech.geomesa.arrow.io.{SimpleFeatureArrowFileReader, SimpleFeatureArrowIO}
+import org.locationtech.geomesa.arrow.vector.{ArrowDictionary, SimpleFeatureVector}
+import org.locationtech.geomesa.utils.io.WithClose
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.filter.Filter
+
+import scala.collection.mutable.ArrayBuffer
+
+class CachingSimpleFeatureArrowFileReader(is: InputStream)(implicit allocator: BufferAllocator)
+    extends SimpleFeatureArrowFileReader  {
+
+  private val opened = ArrayBuffer.empty[CachingSingleFileReader]
+  private val readers = createReaders()
+  private val sort = SimpleFeatureArrowIO.getSortFromMetadata(readers.head.metadata)
+
+  override def sft: SimpleFeatureType = readers.head.sft
+
+  override def dictionaries: Map[String, ArrowDictionary] = readers.head.dictionaries
+
+  override def features(filter: Filter): Iterator[ArrowSimpleFeature] with Closeable = {
+    val skip = new SkipIndicator
+    val nextBatch = SimpleFeatureArrowFileReader.features(sft, filter, skip, sort, dictionaries)
+    new Iterator[ArrowSimpleFeature] with Closeable {
+      // short-circuit if skip is toggled
+      val iter = readers.iterator.takeWhile(_ => !skip.skip).flatMap(_.features(nextBatch, skip))
```

Comment:
although there is some trickiness with the fact that if it's the first time being read, we need to consume the input file iterator up to the next logical file, or the arrow parsing gets out of sync